empire/pkg/runner/runner.go

```diff
@@ -0,0 +1,152 @@
+// package runner provides a simple interface for running docker containers.
+package runner
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"code.google.com/p/go-uuid/uuid"
+
+	"github.com/fsouza/go-dockerclient"
+	"github.com/remind101/empire/empire/pkg/dockerutil"
+	"github.com/remind101/empire/empire/pkg/image"
+	"golang.org/x/net/context"
+)
+
+// DefaultStopTimeout is the number of seconds to wait when stopping a
+// container.
+const DefaultStopTimeout = 10
+
+// RunOpts is used when running.
+type RunOpts struct {
+	// Image is the image to run.
+	Image image.Image
+
+	// Command is the command to run.
+	Command string
+
+	// Environment variables to set.
+	Env map[string]string
+
+	// Streams fo Stdout, Stderr and Stdin.
+	Input  io.Reader
+	Output io.Writer
+}
+
+// Runner is a service for running containers.
+type Runner struct {
+	client *dockerutil.Client
+}
+
+// NewRunner returns a new Runner instance using the docker.Client as the docker
+// client.
+func NewRunner(client *dockerutil.Client) *Runner {
+	return &Runner{client: client}
+}
+
+func (r *Runner) Run(ctx context.Context, opts RunOpts) error {
+	if err := r.pull(opts.Image, opts.Output); err != nil {
+		return fmt.Errorf("runner: pull: %v", err)
+	}
+
+	c, err := r.createContainer(opts)
+	if err != nil {
+		return fmt.Errorf("runner: create container: %v", err)
+	}
+	defer r.removeContainer(c.ID)
+
+	if err := r.startContainer(c.ID); err != nil {
+		return fmt.Errorf("runner: start containeer: %v", err)
+	}
+
+	if err := r.attach(c.ID, opts.Input, opts.Output); err != nil {
+		return fmt.Errorf("runner: attach: %v", err)
+	}
+
+	if err := r.wait(c.ID); err != nil {
+		return fmt.Errorf("runner: wait: %v", err)
+	}
+
+	if err := r.stop(c.ID); err != nil {
+		if _, ok := err.(*docker.ContainerNotRunning); ok {
+			return nil
+		}
+
+		return fmt.Errorf("runner: stop: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Runner) pull(img image.Image, out io.Writer) error {
+	return r.client.PullImage(docker.PullImageOptions{
+		Registry:     img.Registry,
+		Repository:   img.Repository,
+		Tag:          img.Tag,
+		OutputStream: out,
```

Comment:
We probably shouldn't actually be writing this to the connection. This could prevent usage of `emp run` for things like this:

```console
$ emp run cat file -a app > file
```

It would be ideal if we could separate stdout and stderr and write this to stderr.